torchbenchmark/models/nvidia_deeprecommender/nvtrain.py

```diff
@@ -0,0 +1,419 @@
+    elif args.optimizer == "rmsprop":
+      self.optimizer = optim.RMSprop(self.rencoder.parameters(),
+                                lr=self.args.lr, momentum=0.9,
+                                weight_decay=self.args.weight_decay)
+    else:
+      raise  ValueError('Unknown optimizer kind')
+  
+    self.t_loss = 0.0
+    self.t_loss_denom = 0.0
+    self.denom = 0.0
+    self.total_epoch_loss = 0.0
+    self.global_step = 0
+  
+    if self.args.noise_prob > 0.0:
+      self.dp = nn.Dropout(p=self.args.noise_prob)
+
+    if self.toytest:
+      self.toyinputs = torch.randn(128,15178).to(device)
+
+  def DoTrain(self):
+  
+    self.rencoder.train()
+    #if self.args.optimizer == "momentum":
+    #  self.scheduler.step()
+  
+    for i, mb in enumerate(self.data_layer.iterate_one_epoch()):
+  
+      inputs = Variable(mb.cuda().to_dense() if self.args.use_cuda else mb.to_dense())
+
+      self.optimizer.zero_grad()
+  
+      outputs = self.rencoder(inputs)
+  
+      loss, num_ratings = model.MSEloss(outputs, inputs)
+      loss = loss / num_ratings
+      loss.backward()
+      self.optimizer.step()
+      self.global_step += 1
+      self.t_loss += loss.item()
+      self.t_loss_denom += 1
+  
+      if not self.args.nooutput:
+        if i % self.args.summary_frequency == 0:
+          print('[%d, %5d] RMSE: %.7f' % (self.epoch, i, sqrt(self.t_loss / self.t_loss_denom)))
+          self.logger.scalar_summary("Training_RMSE", sqrt(self.t_loss/self.t_loss_denom), self.global_step)
+          self.t_loss = 0
+          self.t_loss_denom = 0.0
+          log_var_and_grad_summaries(self.logger, self.rencoder.encode_w, self.global_step, "Encode_W")
+          log_var_and_grad_summaries(self.logger, self.rencoder.encode_b, self.global_step, "Encode_b")
+          if not self.rencoder.is_constrained:
+            log_var_and_grad_summaries(self.logger, self.rencoder.decode_w, self.global_step, "Decode_W")
+          log_var_and_grad_summaries(self.logger, self.rencoder.decode_b, self.global_step, "Decode_b")
+  
+      self.total_epoch_loss += loss.item()
+      self.denom += 1
+  
+      #if args.aug_step > 0 and i % args.aug_step == 0 and i > 0:
+      if self.args.aug_step > 0:
+        # Magic data augmentation trick happen here
+        for t in range(self.args.aug_step):
+          inputs = Variable(outputs.data)
+          if self.args.noise_prob > 0.0:
+            inputs = dp(inputs)
+          self.optimizer.zero_grad()
+          outputs = self.rencoder(inputs)
+          loss, num_ratings = model.MSEloss(outputs, inputs)
+          loss = loss / num_ratings
+          loss.backward()
+          self.optimizer.step()
+  
+  def train(self, niter=1) :
+    for self.epoch in range(niter):
+
+      if self.toytest:
+        self.rencoder.train()
+        self.optimizer.zero_grad()  
+        outputs = self.rencoder(self.toyinputs)
+        return
```

Comment:
The toytest runs only one iteration no matter what we set for `niter`, could you please fix that?
Also, I see the `DoTrain()` function runs the `model.MSEloss()`, `loss.backward()`, and `self.optimizer.step()` for training, but not in the toytest, could you please add them back?

Reply:
Yikes.